### ISSUE TITLE
add "use items" menu item to crafting table [e]xamine menu

### ISSUE THREAD

**Issue by joe9000** [CleverRaven/Cataclysm-DDA#30243]
problem: currently if you want to [a]pply a pot/pan/etc to heat up frozen food and the pot isn't in your inventory, the [a]pply menu lists all [a]pplyable items within the 3x3 squares centered on where your character is standing. frequently means there isn't a useful key assigned to the pot.

solution: add a "use items" menu item to the crafting table [e]xamine menu that performs an [a]pply action but filters the items to items that are on the crafting table you [e]xamined. now you can put your pot/heating tool on the table you use to craft food and only have three or four items to choose from rather than everything around you that can be [a]pplied.

a further enhancement would be to dynamically add menu items to the crafting menu that represent the [a]ppliable qualities of the items on the crafting table.

adding use item menu:
<img width="500" alt="add use items" src="https://user-images.githubusercontent.com/23408748/57186064-72624e00-6e8d-11e9-9017-65b9eab43f3c.png">

add dynamic menu items (assume there's a pot on the table):
<img width="500" alt="dynamic menu" src="https://user-images.githubusercontent.com/23408748/57186067-77270200-6e8d-11e9-96c8-90dcc76845f4.png">

**Comment by JeanLucVanDamme** [comment]
This seems like a problem worth addressing, but I'd vote against this implementation - it adds even more to the already-overloaded examine menu (which other contributors are actively working to reduce, rather than expand).

What do you think about this instead? Since we now have the ability to mark any item we please as a favorite... Just add a favorites section in the [a]pply menu. Like this:

![CDDA - apply for food items - solve with favorites section](https://user-images.githubusercontent.com/50250164/57186895-997c4800-6eac-11e9-8a27-35fecd9c3095.jpg)

I like this as a general-purpose solution. Whatever the item(s) are that you in particular use repeatedly, you can define them, for easier access within the [a]pply menu. But for your use-case, it needs to be extended to address items not in your inventory... hmmm.

Or, if this annoyance is common, what are the trade-offs of adding a new hotkey? (What do the actual coders think?)

Maybe something like, "activate coo[k]ware". It's just a heavily-filtered apply menu, really. It would only list cooking/defrosting tools. (I know, I know, k and every other letter are taken already. It'd likely have to be unbound by default).

**Comment by JeanLucVanDamme** [comment]
Actually, does anyone think the whole problem is that "TOOLS" is too broad a category? 

Either by-default, or toggled with user intervention, some subcategories within "TOOLS" could be pulled out into their own categories. That'd include COOKWARE. 

I'd find that useful. Even better if you can customize - choose which subcategories are especially important to you, and toggle on/off pulling those out from inside TOOLS.

(You could achieve the same goal by manually favorite-ing the items in that subcategory, assuming [a]pply now has a FAVORITES category, but that's the long way to do it.)

**Comment by esotericist** [comment]
Speaking for myself as a player (rather than as a contributor), I am *really* excited by the proposal in this issue of being able to add tools to specific workstations so that I can spatially organize my activatable items while still keeping them close together.

This is the kind of thing that helps me reason about the space my character is in, and helps me feel that a workshop is an actual *workshop*, rather than merely a collection of tiles which may contain the things a workshop should contain.

As someone who spends a lot of time thinking about UX needs, one interesting thing about this proposal is that implementing it is actually one of the cases where the behavior could be a) optional, b) enabled by default, and c) *still* not actually substantially *impact* anyone in most cases unless they take the extra step of leaving tools on a table/workbench.

Unless the use case in question where it is impactful is "I want to examine in order to get items without anything else in the way", in which case that ship has basically already sailed.

**Comment by JeanLucVanDamme** [comment]
@esotericist - Great Scott, you're right! This proposal has a lot of gameplay and UX potential that I overlooked. 

(Sorry, @joe9000, I went way too deep into only the least-interesting use-case of your proposal.)

Potential menu bloat still seems like a problem - the atmosphere of the game (and the true gameplay) is interrupted to the extent menus become larger, more frequent, and more expansive. When I explore the awesome new features built on this proposal, I don't want it to be through a big menu. That trend takes away from the atmosphere in a lot of ways - a large menu physically blocks more of the lovely tile graphics, takes longer to navigate (more keypresses), etc. ANYWAY, so do everything you guys said, but let me get _straight_ to it with a hotkey, or without menus.

Maybe a directional [a]pply function (just like [g]et now has) accomplishes this. [a]pply towards a specific workbench tile to apply only the tools on that space.

P.S. And these features are _especially_ optional if we tweak how workbenches are designated. So, don't automatically designate every flat surface as a workbench -- only certain surfaces based on sensible, default but customizable rules. Users can manually designate more.

**Comment by joe9000** [comment]
i think there's some real meat to @esotericist's thoughts above... the potential for building a custom workstation is really intriguing. the best way i can articulate it is in terms of dwarf fortress. tortured analogy: building a workshop (e.g., jeweler's workshop) explicitly filters tasks available in the [e]xamine menu to cut and encrust. i assign specific stockpiles to the workshop to explicitly filter the components available to the workshop (e.g., a furniture stockpile limited to masterwork or better and a cut gem stockpile limited to masterwork cut gems). when i [e]xamine the workshop, my options are effectively encrust, because only the encrust task has the required materials. 

translating that to cdda, i designate a square containing a table and a pot as a crafting station; i designate three zones: perishable food (fridge), non-perishable food(display rack), and a heat source (brazier); i assign the three zones to the designated crafting station. [e]xamining the designated crafting station brings up the current menu items: craft items, recraft last recipe, craft as long as possible, and work on craft. appended after work on craft are the [a]ppliable qualities of items in the designated crafting square. for the first four menu options, recipes are filtered to recipes that can be crafted with the tools in the designated crafting station and assigned zones. recipes that could be crafted but are missing ingredients within the assigned zones are grayed out. for the item specific qualities appended the [e]xamine menu, the subsequent menu is an [a]pply menu filtered to items you can heat up, further filtered by the zones assigned to the designated crafting station. 

i really like the idea that an initial hard filter is applied based on the tools placed in your station, and subsequent soft filter based on whether you've got the ingredients readily available. @esotericist's spot on: spatial organization has a natural feel to it.

**Comment by joe9000** [comment]
sorry, still learning git, thought it was going to close back to the issue list :D

**Comment by JeanLucVanDamme** [comment]
I like everything about your last comment.

Bonus, there's tons of room for this to interact with traits, skills, and skill progression.

**Comment by joe9000** [comment]
random thought arising from my evident desire to turn cdda into df2: assigning NPCs to crafting stations. [j][m][q] deluxe oatmeal...

**Comment by esotericist** [comment]
> random thought arising from my evident desire to turn cdda into df2: assigning NPCs to crafting stations. [j][m][q] deluxe oatmeal...

There is a great deal of future functionality for NPCs that we all desperately want.

**Comment by Brambor** [comment]
Another way to achieve heating food is:
1. `a` Use item.
   - A big menu opens:
     ![Image](https://github.com/user-attachments/assets/92b3f6d4-44ae-40d1-bda6-5c64226d3c43)
2. `/` filter `pot` `ENTER` to confirm filter.
    - Only pots are filtered:
      ![Image](https://github.com/user-attachments/assets/adc8b8b2-a0ce-4f81-aad8-b6b69e81047f)
3. Select the pot.

tl;dr: Do `a/pot` `ENTER` `ENTER`
